pkg/kubelet/podcertificate/podcertificate_manager.go

```diff
@@ -0,0 +1,561 @@
+		return
+	}
+
+	klog.InfoS("Checked state of PodCertificateRequest", "state", state)
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	// Clear the pending credential record and make a live credential
+	// record.
+	pendingRecord := m.pendingCredentials[key]
+	delete(m.pendingCredentials, key)
+
+	if pendingRecord == nil {
+		// If we don't have a pending record corresponding to this PCR, there
+		// are two possible causes:
+		//
+		// 1) Some admin manually created a PCR that targets a pod on our node.
+		//
+		// 2) Kubelet created this PCR, but then restarted.
+		//
+		// In both cases, just do nothing.
+		klog.InfoS("PodCertificateRequest update was for stray", "pcr", key)
+		return
+	}
+
+	finalKey := credentialKey{
+		podUID:     newPCR.Spec.PodUID,
+		volumeName: pendingRecord.volumeName,
+		path:       pendingRecord.path,
+	}
+	if state == certificatesv1alpha1.PodCertificateRequestConditionTypeIssued {
+		klog.InfoS("Recording valid credential entry", "poduid", newPCR.Spec.PodUID, "volumename", pendingRecord.volumeName, "path", pendingRecord.path)
+		m.validCredentials[finalKey] = &validCredentialRecord{
+			key:            pendingRecord.privateKey,
+			certificates:   []byte(newPCR.Status.CertificateChain),
+			beginRefreshAt: newPCR.Status.BeginRefreshAt.Time,
+			expiresAt:      newPCR.Status.NotAfter.Time,
+		}
+	} else {
+		klog.InfoS("Recording invalid credential entry", "poduid", newPCR.Spec.PodUID, "volumename", pendingRecord.volumeName, "path", pendingRecord.path)
+		m.invalidCredentials[finalKey] = &invalidCredentialRecord{
+			condition: state,
+			reason:    reason,
+			message:   message,
+		}
+	}
+}
+
+// On delete, clean up the pending credential record corresponding
+// to this PodCertificateRequest, if any exists.
+func (m *IssuingManager) handlePCRDelete(old any) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(old)
+	if err != nil {
+		return
+	}
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	delete(m.pendingCredentials, key)
+}
+
+// createPCR creates a PodCertificateRequest and records it in the
+// pendingCredentials map.  The PodCertificateRequest informer loop will then
+// wait for it to be issued, denied, or failed, and update liveCredentials.
+func (m *IssuingManager) createPCR(
+	ctx context.Context,
+	namespace, podName string, podUID types.UID, serviceAccountName string, serviceAccountUID types.UID, nodeName types.NodeName, nodeUID types.UID, volumeName, path, signerName, keyType string) error {
+
+	privateKey, publicKey, proof, err := m.generateKeyAndProof(keyType, []byte(podUID))
+	if err != nil {
+		return fmt.Errorf("while generating keypair: %w", err)
+	}
+
+	pkixPublicKey, err := x509.MarshalPKIXPublicKey(publicKey)
+	if err != nil {
+		return fmt.Errorf("while marshaling public key: %w", err)
+	}
+
+	keyPEM, err := pemEncodeKey(privateKey)
+	if err != nil {
+		return fmt.Errorf("while PEM-encoding private key: %w", err)
+	}
+
+	req := &certificatesv1alpha1.PodCertificateRequest{
```

Comment:
Add owner reference to the pod so we clean up automatically.  And test with the OwnerReferencesPermissionEnforcement admission plugin.  You may have to add the deletion powers back, but that seems worth the tradeoff.  This prevents build up of podcertificaterequests when pods are short lived.